### PR TITLE
feat: notify users when previously-alerted disruptions clear (#361)

### LINKED ISSUE #361: Alert if disruption clears during notification window
Currently we only alert on changes to disruption during the notification window but suppress "good service" alerts. If during this window we've previously notified of a disruption we should notify when that clears so users know they can ignore the previous alert.

### PR DESCRIPTION
**Pull request by mnbf9rca**
## Summary

Implements issue #361 - Users now receive notifications when previously-alerted disruptions clear, so they can ignore earlier alerts and make informed travel decisions.

### Key Features

- **Status update notifications** sent when lines return to "Good Service" or "No Issues"
  - Email notifications with green "Service Restored" theme
  - SMS notifications with concise format  
  - Both include cleared lines + still-disrupted lines for complete context

- **Smart state tracking** distinguishes cleared vs suppressed states
  - Lines transitioning to cleared states (Good Service, No Issues) → send notification
  - Lines transitioning to suppressed states (e.g., Service Closed) → keep tracking, don't notify yet
  - Cleared lines removed from Redis state after notification sent

- **Future-proofed for #214** via `is_cleared_state` column on `alert_disabled_severities`
  - Cleared states: disruption is over (Good Service, No Issues)
  - Suppressed states: still problematic, just not alerting (Service Closed when implemented)

### Technical Implementation

**Database:**
- Added `is_cleared_state` boolean column to `alert_disabled_severities` table
- Migration sets Good Service (severity 10) and No Issues (severity 18) as cleared states

**Core Logic:**
- New `detect_cleared_lines()` pure function identifies lines that have transitioned to cleared states
- `_process_single_route()` now detects cleared lines and sends status update notifications
- Redis state management updated to remove cleared lines after notification

**Notifications:**
- `NotificationService.send_status_update_email()` - green-themed HTML email
- `NotificationService.send_status_update_sms()` - concise SMS with 160 char limit handling
- New `status_update.html` template

**Testing:**
- 13 new tests covering email/SMS success, errors, Redis state management, edge cases
- All 1578 tests passing
- 95.08% coverage (exceeds 95% threshold)

### Files Changed

- `backend/app/models/tfl.py` - Added `is_cleared_state` column
- `backend/alembic/versions/` - New migration
- `backend/app/schemas/tfl.py` - Added `ClearedLineInfo` schema
- `backend/app/services/alert_service.py` - Detection, notification, state management (+318 lines)
- `backend/app/services/notification_service.py` - Status update methods
- `backend/app/templates/email/status_update.html` - New template
- `backend/app/core/redis.py` - Added `delete()` to RedisClientProtocol
- `backend/tests/test_alert_service.py` - 13 new tests
- `backend/tests/test_notification_service.py` - Updated imports

### Example Notification

**Email Subject:** "Service Restored - [Route Name]"

**Email Body:**
> ✅ Service Restored:
> Victoria - Good Service (Previously: Severe Delays)
>
> ⚠️ Still Disrupted:
> Piccadilly - Minor Delays

This gives users complete context to make informed decisions.

## Test Plan

- [x] All 1578 tests passing
- [x] 95.08% code coverage (exceeds 95% threshold)
- [x] All pre-commit hooks passing (ruff, mypy, etc.)
- [x] Cleared line detection works correctly
- [x] Status update emails sent with correct content
- [x] Status update SMS sent with correct format
- [x] Redis state management removes cleared lines
- [x] Suppressed states (non-cleared) kept in Redis for future tracking

Closes #361

## Summary by Sourcery

Add cleared-disruption detection and status update notifications so users are informed when previously alerted lines return to normal service.

New Features:
- Introduce cleared line tracking via ClearedLineInfo and detection logic to identify previously disrupted lines that reach cleared states.
- Send status update emails and SMS messages summarizing cleared lines and remaining disruptions using a new green-themed email template.
- Track cleared vs suppressed severities using an is_cleared_state flag on alert_disabled_severities, with a migration seeding Good Service and No Issues states.

Enhancements:
- Extend alert processing to fetch both filtered and unfiltered disruptions, use stored Redis state to compute cleared lines, and clean up alert state after status updates.
- Improve alert cooldown and state handling by returning stored line state from _should_send_alert and adding robust Redis state update logic for cleared lines.
- Augment Redis client protocol with delete support and expand alert logging helpers and tests for better resilience and observability.

Tests:
- Add comprehensive unit and integration tests for cleared line detection, cleared state retrieval, status update sending, Redis state cleanup, and error handling in alert and notification services.

### CODE REVIEW

File: backend/app/services/alert_service.py [hunk 1]
```diff
@@ -1271,47 +1720,87 @@ async def _process_single_route(
...
+                            route_name=route.name,
+                            cleared_count=len(cleared_lines),
+                            cleared_line_ids=[cl.line_id for cl in cleared_lines],
+                        )
+                        # Send status update notifications
+                        await self._send_status_update_notifications(
+                            route=route,
+                            schedule=active_schedule,
+                            cleared_lines=cleared_lines,
+                            still_disrupted=disruptions,  # Current alertable disruptions
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/alert_service.py]: Thank you for the suggestion. However, the current behavior is intentional and correct.

Passing filtered `disruptions` (alertable severities only) as `still_disrupted` is semantically correct because:

1. Users should only see disruptions they were/would be alerted about
2. Showing non-alertable severities like "Service Closed" as "still disrupted" would confuse users who weren't alerted about them in the first place
3. This maintains consistency with the alert system's filtering logic

The status update notification's purpose is to inform users about lines that have cleared from previously alerted states, and to provide context about what's still alertable - not to show all TfL disruptions regardless of severity.

**Comment by mnbf9rca** [comment]
## PR Review Comments Addressed

All review comments have been addressed. Summary:

### ✅ Accepted Changes (4 comments):

1. **Jinja template fix** (PRRT_kwDOQNO0Es5lOxH9) - Changed to `|default()` filter ✅
2. **SMS truncation bug** (PRRT_kwDOQNO0Es5lONzR) - Added multi-stage truncation ✅
3. **Cache cleared states** (Sourcery) - Eliminated per-route DB queries ✅
4. **Integration test** (PRRT_kwDOQNO0Es5lONzI) - Added all-lines-clearing test ✅

### ❌ Rejected Changes (2 comments):
- **still_disrupted filtering** - Intentional design (replied with rationale)
- **Migration ON CONFLICT** - Correct for idempotent migrations

### Test Results:
- ✅ All 1591 tests passing
- ✅ 95.68% coverage (exceeds 95% requirement)
- ✅ All pre-commit hooks passing

Commit: db080a7

File: backend/tests/test_alert_service.py [hunk 2]
```diff
@@ -5018,3 +5267,501 @@ def now(cls, tz: None | type[UTC] = None) -> datetime:  # type: ignore[override]
...
+            schedule=schedule,
+            cleared_lines=sample_cleared_lines,
+            still_disrupted=sample_disruptions,
+        )
+
+        # Verify no notification logs were created
+        result = await db_session.execute(select(NotificationLog).where(NotificationLog.route_id == route.id))
+        logs = result.scalars().all()
+        assert len(logs) == 0
+
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/test_alert_service.py:5767]: Fixed in db080a7.

Added integration test `test_all_lines_cleared_sends_status_update` that verifies status updates are sent when all lines return to Good Service (disruptions list empty but all_route_disruptions contains Good Service entries).

The test confirms the full flow works correctly even when the alertable disruptions list is empty.

File: backend/app/services/notification_service.py [hunk 3]
```diff
@@ -203,6 +203,182 @@ async def send_disruption_sms(
...
+
+                # Add URL
+                url = " More: tfl.gov.uk/tube-dlr-overground/status/"
+
+                # Build final message
+                message = f"{base_msg}{disrupted_msg}{url}"
+
+                # If too long, drop the "still disrupted" part
+                if len(message) > SMS_MAX_LENGTH:
+                    message = f"{base_msg}{url}"
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/notification_service.py:356]: Fixed in db080a7.

Implemented multi-stage truncation to handle all edge cases:
1. Drop "still disrupted" section
2. Reduce cleared lines from 3 to 2
3. Reduce cleared lines from 2 to 1
4. Truncate route_name and line names with ellipsis
5. Force truncate to 160 chars as final fallback

Added 4 comprehensive tests covering each truncation stage.

File: backend/app/templates/email/status_update.html [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+</head>
+<body>
+    <div class="container">
+        <div class="header">
+            <div class="alert-icon">✅</div>
+            <h1>Service Restored</h1>
+            <div class="route-name">{{ route_name }}</div>
+        </div>
+        <div class="content">
+            <p class="greeting">Hello {{ user_name or "there" }},</p>
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/templates/email/status_update.html]: Fixed in db080a7.

Changed from `{{ user_name or "there" }}` to `{{ user_name|default("there") }}` to properly handle empty strings (not just None values).

**Comment by mnbf9rca** [comment]
**Additional improvement:** Updated migration to use `ON CONFLICT DO UPDATE` instead of `DO NOTHING` (commit 045905e).

This ensures any pre-existing severity 18 records with `is_cleared_state=false` are corrected to `true`, maintaining data consistency after migration runs.
